### DISCUSSION TITLE
Perturbation in FastGradientMethod

### DISCUSSION THREAD

**Discussion by umbertogagl97** [Trusted-AI/adversarial-robustness-toolbox#1314] [Q&A] [answered]
Hello,
I am new to ART and have a question about the Fast Gradient Sign attack [https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/art/attacks/evasion/fast_gradient.py](url).
Is there a fast way to obtain the perturbation applied to every image? I'm tring the subtraction adversarial_image - original_image but I have problem with values of the result because they're in [-1,1] and I need to normalize them in [0,1]. Thank you.
I report the code:
`attack = FastGradientMethod(estimator=classifier, eps=0.1)`
`x_test_adv = attack.generate(x=x_test,y=y_test)`
`perturb=x_test_adv-x_test`

**Comment by beat-buesser** [comment, marked as answer]
Hi @umbertogagl97 Thank you very much for your interest in ART! What is the pixel range of `x_test` and how do you define `classifier`?

**Reply by umbertogagl97** [reply]
Hi,
x_test is in [0,1] and classifier is defined like that:
`classifier = KerasClassifier(model=model, clip_values=(min_, max_))`
where min_=0, max_=1 and model is a pre-trained network (vgg16) with fine tuning.
The train set is also in [0,1].

**Reply by beat-buesser** [reply]
Ok, why is your result in range [-1, 1]?

**Reply by umbertogagl97** [reply]
Because original and adversarial images are in [0,1] but the adversarial images can have some smaller pixels and the subtraction returns negative values. For this reason I have thought about normalization:
`min_p, max_p = np.amin(perturb), np.amax(perturb)`
`perturb = (perturb - min_p) / (max_p - min_p)`
Is there a faster way to use fast_gradient.py to obtain the perturbation or something similar?

**Reply by beat-buesser** [reply]
This should not be necessary, the adversarial image should not have values outside of `clip_values=(min_, max_)`. Do you have a complete example that we can run?

**Reply by umbertogagl97** [reply]
yes, I have a notebook on colab.
I Insert the code here
[Attacco_VGG16_keras_FGM_stl10.zip](https://github.com/Trusted-AI/adversarial-robustness-toolbox/files/7171868/Attacco_VGG16_keras_FGM_stl10.zip)
Or the file .py :
[attacco_vgg16_keras_fgm_stl10 (2).zip](https://github.com/Trusted-AI/adversarial-robustness-toolbox/files/7171884/attacco_vgg16_keras_fgm_stl10.2.zip)

**Reply by umbertogagl97** [reply]
I try to explain the problem better. The adversarial images are not the problem, they are in [0,1], but the perturbation (calculated by subtraction) contains value <0. For this reason I would like to know if there is another way to get the perturbation.

**Reply by beat-buesser** [reply]
Ok, that makes sense. In this case, the adversarial perturbation can be negative, this happens in cases where a pixel had a negative loss gradient and therefore its pixel value decreased.

**Reply by umbertogagl97** [reply]
Okay. So, is the subtraction the right and only way to calculate the perturbation?

**Reply by beat-buesser** [reply]
Yes, if you would like to know how much the pixel values have changed. Why do you ask?

**Reply by umbertogagl97** [reply]
Because I'm just start to working with ART and I don't know yet how to use well the library. Thank you very much.
